src/Shared/NodeEndpointOutOfProcBase.cs

```diff
@@ -374,77 +365,85 @@ private void PacketPumpProc()
-                    // respond with another long.  Once the handshake is complete, both sides can be assured the
-                    // other is ready to accept data.
-                    // To avoid mixing client and server builds, the long is the MSBuild binary timestamp.
-
-                    // Compatibility issue here.
-                    // Previous builds of MSBuild 4.0 would exchange just a byte.
-                    // Host would send either 0x5F or 0x60 depending on whether it was the toolset or not respectively.
-                    // Client would return either 0xF5 or 0x06 respectively.
-                    // Therefore an old host on a machine with new clients running will hang, 
-                    // sending a byte and waiting for a byte until it eventually times out;
-                    // because the new client will want 7 more bytes before it returns anything.
-                    // The other way around is not a problem, because the old client would immediately return the (wrong)
-                    // byte on receiving the first byte of the long sent by the new host, and the new host would disconnect.
-                    // To avoid the hang, special case here:
-                    // Make sure our handshakes always start with 00.
-                    // If we received ONLY one byte AND it's 0x5F or 0x60, return 0xFF (it doesn't matter what as long as
-                    // it will cause the host to reject us; new hosts expect 00 and old hosts expect F5 or 06).
+                    // The handshake protocol is a series of int exchanges.  The host sends us a each component, and we
+                    // verify it. Afterwards, the host sends an "End of Handshake" signal, to which we respond in kind.
+                    // Once the handshake is complete, both sides can be assured the other is ready to accept data.
+                    Handshake handshake = GetHandshake();
                     try
                     {
-                        long handshake = localReadPipe.ReadLongForHandshake(/* reject these leads */ new byte[] { 0x5F, 0x60 }, 0xFF /* this will disconnect the host; it expects leading 00 or F5 or 06 */
+                        int index = 1;
+                        foreach (int part in handshake.RetrieveHandshakeComponents())
+                        {
+
+                            int handshakePart = _pipeServer.ReadIntForHandshake(index == 1 ? 0x01 : 0x00 /* this will disconnect a < 16.8 host; it expects leading 00 or F5 or 06. 0x00 is a wildcard */
 #if NETCOREAPP2_1 || MONO
                             , ClientConnectTimeout /* wait a long time for the handshake from this side */
 #endif
                             );
 
-#if FEATURE_SECURITY_PERMISSIONS
-                        WindowsIdentity currentIdentity = WindowsIdentity.GetCurrent();
-#endif
+                            if (handshakePart != part)
```

Comment:
Is there any chance the various ways how a handshake can fail can be tested with a unit test? Client sends a shorter handshake and it will be detected here, client sends a longer handshake and it will be detected in `ReadEndOfHandshakeSignal` below, etc.

Reply:
I think...maybe? The hardest part, I think, would be mocking a pipe. If I can do that, I don't actually have to make handshakes; I can just send my customized ints and see when they disconnect/throw errors.

Of course, that only tests part of what really happens with a handshake. It would be ideal if I can call TryConnectToProcess with a MockProcess, but that sounds like it would be...complicated.

Reply:
Got it, thanks. I guess it would be nice if the logic of de/serializing the handshake and detecting a match was all abstracted away into its own helper class that would be unit-testable without elaborate mocking but I understand that it's not super easy. It's just a bit unfortunate that the only change made to test code in this PR is a test case deletion.